### ISSUE TITLE
Query for date range intersection (ie: date range overlapping with another date range)

### ISSUE THREAD

**Issue by feliperaul** [ankane/searchkick#973]
I'm having issues trying to use Searchkick for returning records with date ranges overlapping (intersection) with the search form date range. 

In other words: My model has two `Date` columns: a `start_date` and `end_date`. And the search form has a `period_start` and `period_end`. I must then return all records that have an intersection between these two dates.

Using ActiveRecord, the query I've been using for years is like this:

`Model.where("DATEDIFF(start_date, ?) * DATEDIFF(?, end_date)) >= 0", period_end, period_start")` 
(Source: https://makandracards.com/makandra/984-test-if-two-date-ranges-overlap-in-ruby-or-rails)

I'm not being able to recreate this in Searchkick, and I have no clue in how to do it. 

I could only find this on the interwebs: 
https://stackoverflow.com/questions/16797694/elasticsearch-date-range-intersection

But the accepted solution uses ElasticSearch's `range` query, and the only thing I can find on `range` queries on the documentation are deprecated facets and aggregations, which I assume have nothing to do with i'm trying to achieve.

**Comment by mikelkew** [comment]
Not sure if there is a range intersection feature in ES, there may be.

Failing that, another approach is to create a custom field containing an array of all the dates in the range, and then do the same when applying the filter.

Here's a rough example:
```
# custom field
dates: (start_date..end_date).map(&:to_s)

# query
Model.search(where: {dates: (Date.parse(period_start)..Date.parse(period_end)).map(&:to_s)})
```

**Comment by feliperaul** [comment]
@mikelkew Unfortunately my date ranges are very long, that would make for huge arrays. 

I'm posting my current hacky fix so it helps someone later on: If the date range fields are filled in, I use ActiveRecord to query for the records that match the date intersection, and `.pluck(:id)`, and I also add this condition to the Searchkick hash: `where_hash.merge!({id: scope.ids})`

It's on extra DB query I'd like to avoid, but it solved the problem.

**Comment by ankane** [comment]
Hey @feliperaul, glad you found a solution. As far as I know, there is not an equivalent to the SQL in your first post in Elasticsearch.

**Comment by elainesit** [comment]
I know this is old but I'm thinking through the same thing, with the minor complication that the end_date is not usually present. I wanted to do something clever with ranges, but actually this works fine:

```ruby
filters[:begins_at] = {gte: Date.parse(params[:from_date])}
filters[:ends_at] = {lte: Date.parse(params[:to_date])}
```

with the index:

```
{
      ...
      begins_at: start,
      ends_at: finish.presence || start
}

Should it be more complicated? I don't mind adding range support if it's useful.

**Comment by will-r** [comment]
I know this is old but I'm thinking through the same thing, with the minor complication that the end_date is not usually present. I wanted to do something clever with ranges, but actually this works fine:

```ruby
filters[:begins_at] = {gte: Date.parse(params[:from_date])}
filters[:ends_at] = {lte: Date.parse(params[:to_date])}
```

with the index:

```
{
      ...
      begins_at: start,
      ends_at: finish.presence || start
}

Should it be more complicated? I don't mind adding range support if it's useful.

**Comment by feliperaul** [comment]
@will-r Your comment made me revisit this issue. 

If you open the [Makandra cards link](https://makandracards.com/makandra/984-test-if-two-date-ranges-overlap-in-ruby-or-rails)) I posted when I opened the issue, you'll se it was edited 2 months ago. 

In the most recent version, the suggested query is exactly this:

```
scope :overlapping, -> { |interval|
    where(start_date <= ? AND ? <= end_date", interval.end_date, interval.start_date)
  }
```

Looking more into it (could it be that simple?!), I confirmed that this is, in fact, the algo provided in the most upvoted question in StackOverflow about overlapping date ranges:

> (StartA <= EndB) and (EndA >= StartB)

(source [here](https://stackoverflow.com/a/325964/1290457) if you want to understand why it works).

There's one huge plus to this simpler algorithm: it's WAY faster. In my database, with 3.6 million records, the original SQL I posted sometime takes generally 5 SECONDS to find matches, which made me resort to all sorts of hacks to achieve what I wanted. With the new simpler query, it takes 10ms if there are few results, and 1.5 seconds tops if there are many thousands. 

So, yes, you are totally right. I refactored my code to this:

```
          where_query[:from_date] = { lte: target_interval.to_date }
          where_query[:to_date]   = { gte: target_interval.from_date }
```

It works perfectly!
